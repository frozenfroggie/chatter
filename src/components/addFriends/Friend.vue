<template>
  <li class='conversationLabel' @click='openChat'>
    <div class='avatarContainer'>
      <icon name="user-circle" class='avatar' scale="2.5"></icon>
      <div :class='{online: friend.isOnline}'></div>
    </div>
    <div class='content'>
      <div class='conversationName'>
        {{ friend.name }}
      </div>
    </div>
    <icon name="plus" class='plus'></icon>
  </li>
</template>

<script>
export default {
  props: ['friend'],
  methods: {
    openChat () {
      this.$router.push(this.conversation.name)
    }
  }
}
</script>

<style scoped>
.conversationLabel {
  display: grid;
  grid-template-columns: 0px 50px auto 30px 0px;
  grid-template-rows: 60px;
  grid-column-gap: 5px;
  align-items: center;
  grid-template-areas:
    ". avatar content add ."
}
.conversationLabel:hover {
  cursor: pointer;
}
.avatarContainer {
  position: relative;
  height: 38px;
  width: 38px;
  border-radius: 50%;
  grid-area: avatar;
  justify-self: center;
}
.avatar {
  color: white;
  opacity: .4;
}
.plus {
  color: white;
  opacity: .4;
  grid-area: add;
}
.plus:hover {
  transition: opacity .6s;
  opacity: .9;
}
.online {
  position: absolute;
  bottom: 0px;
  right: 0px;
  height: 6px;
  width: 6px;
  border: 3px solid rgb(46, 59, 79);
  border-radius: 50%;
  background-color: rgb(66, 183, 42);
}
.content {
  grid-area: content;
}
.conversationName {
  text-align: left;
  color: white;
  font-size: 0.9em;
  font-weight: 600;
  opacity: .9;
}
.conversationPreview {
  line-height: 1.5;
  font-size: 0.8em;
  text-align: left;
  color: rgba(250,250,250,0.5);
}
</style>
