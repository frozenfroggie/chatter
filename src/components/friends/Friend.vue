<template>
  <li>
    <div
      @click='openChat'
      :to="{ name: 'chat', params: {conversationId}}"
      class='conversationLabel'
      tag='div'>
      <div class='avatarContainer'>
        <icon name="user-circle" class='avatar' scale="2.5"></icon>
        <div :class='{online: isOnline}'></div>
      </div>
      <div class='content'>
        <div class='friendName'>
          {{ friend.username.charAt(0).toUpperCase() + friend.username.slice(1) }}
        </div>
      </div>
    </div>
  </li>
</template>

<script>
import axiosAuth from '../../config/axiosAuth'

export default {
  props: ['friend', 'conversations'],
  data () {
    return {
      conversationId: undefined
    }
  },
  mounted () {
    axiosAuth.get('/chat/conversationId/' + this.friend._id).then(res => {
      this.conversationId = res.data.conversationId
    }).catch(() => {
      // ignore
    })
  },
  methods: {
    openChat () {
      this.$store.dispatch('getMessages', {
        conversationId: this.conversationId,
        skipMessagesAmount: 0
      })
      this.$router.push('/chat/' + this.conversationId)
    }
  },
  computed: {
    isOnline () {
      const activeConversationsIds = this.$store.getters.activeConversationsIds
      if (!activeConversationsIds) {
        return false
      }
      for (let i = 0; i < activeConversationsIds.length; i++) {
        return activeConversationsIds[i] === this.conversationId
      }
    }
  }
}
</script>

<style scoped>
.conversationLabel {
  display: grid;
  grid-template-columns: 0px 50px auto 30px 0px;
  grid-template-rows: 60px;
  grid-column-gap: 5px;
  align-items: center;
  grid-template-areas:
    ". avatar content timestamp ."
}
.conversationLabel:hover {
  cursor: pointer;
}
.avatarContainer {
  position: relative;
  height: 38px;
  width: 38px;
  border-radius: 50%;
  grid-area: avatar;
  justify-self: center;
}
.avatar {
  color: white;
  opacity: .4;
}
.online {
  position: absolute;
  bottom: 0px;
  right: 0px;
  height: 6px;
  width: 6px;
  border: 3px solid rgb(46, 59, 79);
  border-radius: 50%;
  background-color: rgb(66, 183, 42);
}
.content {
  grid-area: content;
}
.friendName {
  text-align: left;
  color: white;
  font-size: 0.9em;
  font-weight: 600;
  opacity: .9;
}
</style>
