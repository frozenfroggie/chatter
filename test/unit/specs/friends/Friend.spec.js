import { shallowMount, RouterLinkStub } from '@vue/test-utils'
import Friend from '@/components/friends/Friend.vue'

describe('Friend', () => {
  it('renders friend name in div.friendName when props.friend passed', () => {
    const friend = {
      name: 'Christianna Lynn',
      isOnline: true
    }
    const wrapper = shallowMount(Friend, {
      propsData: { friend },
      stubs: {
        icon: true,
        RouterLink: RouterLinkStub
      }
    })
    expect(wrapper.find('div.friendName').text()).toBe(friend.name)
  })
  it('matches snapshot', () => {
    const friend = {
      name: 'Christianna Lynn',
      isOnline: true
    }
    const wrapper = shallowMount(Friend, {
      propsData: { friend },
      stubs: {
        icon: true,
        RouterLink: RouterLinkStub
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
