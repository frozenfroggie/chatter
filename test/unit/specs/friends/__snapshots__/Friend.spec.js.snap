// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Friend matches snapshot 1`] = `
<li class="conversationLabel">
    <div class="avatarContainer">
        <!---->
        <div class="online"></div>
    </div>
    <div class="content">
        <div class="friendName">
            Christianna Lynn
        </div>
    </div>
</li>
`;
